gnmi_server/gnoi_system.go

```diff
@@ -0,0 +1,377 @@
+		resp = &syspb.RebootResponse{}
+	case rebootStatusKey:
+		req = req.(*syspb.RebootStatusRequest)
+		resp = &syspb.RebootStatusResponse{}
+	case rebootCancelKey:
+		req = req.(*syspb.CancelRebootRequest)
+		resp = &syspb.CancelRebootResponse{}
+	}
+
+	reqStr, err := json.Marshal(req)
+	if err != nil {
+		log.V(2).Infof("[Reboot_Log] Error in marshalling JSON: %v", err)
+		return nil, status.Errorf(codes.Internal, err.Error()), msgDataStr
+	}
+	// Publish to notification channel.
+	if err := np.Send(rebootNotifKey, "", map[string]string{dataMsgFld: string(reqStr)}); err != nil {
+		log.V(2).Infof("[Reboot_Log] Error in publishing to notification channel: %v", err)
+		return nil, status.Errorf(codes.Internal, err.Error()), msgDataStr
+	}
+
+	// Wait for response on Reboot_Response_Channel.
+	tc := time.After(notificationTimeout)
+	for {
+		select {
+		case msg := <-channel:
+			op, data, fvs, err := processMsgPayload(msg.Payload)
+			if err != nil {
+				log.V(2).Infof("[Reboot_Log] Error while receiving Response Notification = [%v] for message [%v]", err.Error(), msg)
+				return nil, status.Errorf(codes.Internal, fmt.Sprintf("Error while receiving Response Notification: [%s] for message [%s]", err.Error(), msg)), msgDataStr
+			}
+			log.V(2).Infof("[Reboot_Log] Received on the Reboot notification channel: op = [%v], data = [%v], fvs = [%v]", op, data, fvs)
+
+			if op != rebootNotifKey {
+				log.V(2).Infof("[Reboot_Log] Op: [%v] doesn't match for `%v`!", op, rebootNotifKey)
+				continue
+			}
+			if fvs != nil {
+				if _, ok := fvs[dataMsgFld]; ok {
+					msgDataStr = fvs[dataMsgFld]
+				}
+			}
+			if swssCode := swssToErrorCode(data); swssCode != codes.OK {
+				log.V(2).Infof("[Reboot_Log] Response Notification returned SWSS Error code: %v, error = %v", swssCode, msgDataStr)
+				return nil, status.Errorf(swssCode, "Response Notification returned SWSS Error code: "+msgDataStr), msgDataStr
+			}
+			return resp, nil, msgDataStr
+
+		case <-tc:
+			// Crossed the reboot response notification timeout.
+			log.V(2).Infof("[Reboot_Log] Response Notification timeout from NSF Manager!")
+			return nil, status.Errorf(codes.Internal, "Response Notification timeout from NSF Manager!"), msgDataStr
+		}
+	}
+}
+
+// Reboot implements the corresponding RPC.
+func (srv *Server) Reboot(ctx context.Context, req *syspb.RebootRequest) (*syspb.RebootResponse, error) {
+	ctx, err := authenticate(srv.config, ctx)
+	if err != nil {
+		return nil, err
+	}
+	log.V(2).Info("gNOI: Reboot")
+	if err := validRebootReq(req); err != nil {
+		return nil, status.Errorf(codes.InvalidArgument, err.Error())
+	}
+
+	// Initialize State DB.
+	rclient, err := getRedisDBClient(stateDB)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, err.Error())
+	}
+	defer rclient.Close()
+
+	// NSF WARM BOOT.
+	if req.GetMethod() == syspb.RebootMethod_NSF {
+		// NSF pre-check validation before sending to NSF Manager.
+		return precheckNSFReboot(ctx, srv, req, rclient)
+	}
+
```

Comment:
What happens for other Reboot methods?

Reply:
For the Reboot Method NSF, we perform additional validations such as checking for ongoing NSF requests, etc. For other Reboot methods, the request is directly sent to the Notification channel to initiate reboot